components/datetime/src/provider/mod.rs

```diff
@@ -193,5 +193,17 @@ pub mod gregory {
             pub medium: Cow<'static, str>,
             pub short: Cow<'static, str>,
         }
+
+        #[derive(Debug, PartialEq, Clone, Default)]
+        #[cfg_attr(
+            feature = "provider_serde",
+            derive(serde::Serialize, serde::Deserialize)
+        )]
+        pub struct DateTimeFormatsV1 {
+            pub style_patterns: StylePatternsV1,
+
+            #[cfg_attr(feature = "provider_serde", serde(with = "tuple_vec_map"))]
+            pub available_formats: Vec<(Cow<'static, str>, Cow<'static, str>)>,
```

Comment:
If you put cows in a data struct, you should add a lifetime parameter to the data struct instead of using `'static`.  Alternatively, you can use TinyStr (for skeletons) or SmallStr (for patterns) and obviate the need for a lifetime parameter.

It looks like the skeletons are always, or almost always, 8 ASCII characters or shorter, in which case TinyStr8 would work very well.  If you want to allow skeletons to be longer than that, consider TinyStrAuto.

Reply:
> If you put cows in a data struct, you should add a lifetime parameter to the data struct instead of using 'static. 

This is where a lot of my confusion over the memory model for data providers really stems. All of the other strings are using the `'static` lifetime parameter, so I'm not sure if there is any difference between that usage, and this usage. I was honestly a bit surprised this code compiled.

```
    symbols!(months, [Cow<'static, str>; 12]);

    symbols!(weekdays, [Cow<'static, str>; 7]);

    symbols!(
        day_periods {
            am: Cow<'static, str>,
            pm: Cow<'static, str>,
            noon: Option<Cow<'static, str>>,
            midnight: Option<Cow<'static, str>>,
        }
    );
```

I investigated the source of the Cows, which appears to be #256. It seems like this was initially landed to get the data in tree, with the idea that we would optimize in the future. Similarly I'm a bit nervous about spending a lot of time optimizing at the beginning, but let's start with a decent choice. I think there are different trade-offs in this data provider format, but I'd be happy to try the the non-heap/inline solutions first, and then measure the memory afterwards.

Should I file an issue to add lifetime parameters to the `Cow` structures that are already here? If we're not adding new ones, it seems like the existing ones should be addressed at some point.

Reply:
Here is a breakdown of pattern lengths across all of the CLDR.

<img width="602" alt="image" src="https://user-images.githubusercontent.com/1588648/107390751-9ec4e880-6abd-11eb-931d-9bd8ecb9cf9f.png">

Reply:
Re-posting this with correct byte size, instead of character count:

<img width="602" alt="image" src="https://user-images.githubusercontent.com/1588648/107393532-873b2f00-6ac0-11eb-9e88-365127232138.png">

Reply:
Per the style guide:

> - If lifetime parameters are allowed, use `&'a str`.
> - If lifetime parameters are not allowed, use one of:
>   - [TinyStr](https://github.com/zbraniecki/tinystr) if the string is ASCII-only.
>   - [SmallString](https://crates.io/crates/smallstr) for shorter strings, with a stack size ∈ {8, 12, 16, 20} that fits at least 99% of cases.
>   - `Cow<'static, str>`  for longer strings.

The largest suggested stack size of 20 for this data set is: (aside: I'm not sure if stack size in the style guide is including the data required by by the SmallString struct itself beyond just the data)

20 bytes: 96.5%
29 bytes: 98.8%
30 bytes: 99.5%

Which, according to the style guide suggests we use `Cow<'static, str>`.

Reply:
I need to update the style guide. As of my recent changes to the data provider trait, cows can and should have real lifetime parameters. I thought I updated all existing structs when I did that PR but I guess I missed this one hiding in a macro. You don't have to worry about migrating in this PR; I'll follow up in #257.

About data types: I'm more talking about skeletons than patterns. I believe that patterns are going to be long enough that you need a cow. However, I think the keys (skeletons) might be a good just case for TinyStr.